### PR TITLE
protocol ListBuilder

### PR DESCRIPTION
**Pull request by RoBo-Inc**
Introducing the `ListBuilder` protocol with a default function implementation can significantly reduce the amount of generated code in the `Sources/SwiftSyntaxBuilder/generated/ResultBuilders.swift` file. Additionally, it makes the codebase more comprehensible by emphasizing similarities among all the result builders and highlighting differences between them.

Apart from this, the PR addresses another issue that requires detailed explanation. Two generated result builders (`ExprListBuilder` and `UnexpectedNodesBuilder`) stood out from the others. They were "special" because their `Expression` was not a concrete type; instead, it was a protocol (`ExprSyntaxProtocol` and `SyntaxProtocol` respectively). So we were dealing with existentials here, which are not very good. Furthermore, we had to create additional extensions dedicated to these two result builders, located at the bottom of the `Sources/SwiftSyntaxBuilder/ResultBuilderExtensions.swift` file, accompanied by a lengthy comment.

The key point is that these two result builders don't need to be treated as special cases, and they don't have to involve existentials. They can conform to the `ListBuilder` protocol, just like all the other builders do. Despite this, we can still make them work with protocols just as before, maintaining the same approach used for some other result builders in the same file (`Sources/SwiftSyntaxBuilder/ResultBuilderExtensions.swift`). This eliminates the need for special treatment and enhances codebase consistency.

In summary, this PR represents a refactoring. All the functionality remains unchanged. However, the refactoring is significant, making the code more structured, consistent, and efficient.

### CODE REVIEW

File: Sources/SwiftSyntaxBuilder/ResultBuilderExtensions.swift [hunk 1]
```diff
@@ -50,20 +50,14 @@ extension MemberBlockItemListBuilder {
...
 
-// MARK: Initializing collections from protocols
-// These initializers allow the creation of syntax collections that have a base
-// node as their element from the corresponding protocol type.
-// These are used by the result builders.
-// Since we only have two of these, it doesn’t make sense to generate them.
-
-extension ExprListSyntax {
-  init(_ elements: [ExprSyntaxProtocol]) {
-    self = ExprListSyntax(elements.map { ExprSyntax(fromProtocol: $0) } as [ExprSyntax])
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/ResultBuilderExtensions.swift:61]: I think we still need this so that you can write eg. 

```swift
ExprListSyntax([DeclReferenceExprSyntax(…), ForceUnwrapExprSyntax(…)])`
```

Or any heterogeneous array of expressions, which is what we do in the tests you changed, I think.
Otherwise this is an API-breaking change.

Same for the initializer on `UnexpectedNodesSyntax` below.

File: Sources/SwiftSyntaxBuilder/generated/ResultBuilders.swift [hunk 2]
```diff
@@ -15,82 +15,10 @@
...
-  /// which defaults to Component if buildExpression() is not provided.
-  public typealias Expression = AccessorDeclSyntax
-  
-  /// The type of a partial result, which will be carried through all of the
-  /// build methods.
-  public typealias Component = [Expression]
-  
+public struct AccessorDeclListBuilder: ListBuilder {
   /// The type of the final returned result, which defaults to Component if
   /// buildFinalResult() is not provided.
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/generated/ResultBuilders.swift:20]: I don’t think theses comments provide much value anymore. I would just remove them.

File: Sources/SwiftSyntaxBuilder/generated/ResultBuilders.swift [hunk 3]
```diff
@@ -188,93 +41,21 @@ public extension ArrayElementListSyntax {
...
-  /// Required by every result builder to build combined results from
-  /// statement blocks.
-  public static func buildBlock(_ components: Self.Component...) -> Self.Component {
-    return components.flatMap {
-      $0
-    }
-  }
-  
   /// If declared, provides contextual type information for statement
   /// expressions to translate them into partial results.
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/generated/ResultBuilders.swift:50]: Similarly here, I don’t think the comments provide much value and I would just remove them.

**Review by ahoppen — COMMENTED** [review]
Removing >3000 LOC. I love it 😍

File: Sources/SwiftSyntaxBuilder/ResultBuilderExtensions.swift (hunk 1 above)
Comment (RoBo-Inc) [reply Sources/SwiftSyntaxBuilder/ResultBuilderExtensions.swift:61]: Yeah, no problem. I will put it back 🙂

**Comment by RoBo-Inc** [comment]
Updated ✅

1. `generated/ResultBuilders` file:
-    Removed the unnecessary comments as requested.
-    Added "MARK: -" separators instead, for better readability.

2. Put the extensions back, but:
- Moved them from `ResultBuilderExtensions` to `ConvenienceInitializers` file. Because these extensions are not specifically related to the result builders anymore. At the same time they indeed are convenience initializers.
- Improved its bodies a bit.
- Made the both initializers to be public for consistency (before only one of them was public).

Note:  The `ConvenienceInitializers` file has warning to insert new extensions in alphabetical order. But I just append the two extensions (which I moved) to the end of the file. Because the extensions are related and has the common comment. Besides the extensions in the file actually are not in alphabetical order anyways... 😳

File: Sources/SwiftSyntaxBuilder/ConvenienceInitializers.swift [hunk 4]
```diff
@@ -380,6 +380,22 @@ extension VariableDeclSyntax {
...
+    self.init(elements.map { ExprSyntax(fromProtocol: $0) } as [ExprSyntax])
+  }
+}
+
+extension UnexpectedNodesSyntax {
+  public init(_ elements: [SyntaxProtocol]) {
+    self.init(elements.map { Syntax(fromProtocol: $0) } as [Syntax])
+  }
+}
+
```
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/ConvenienceInitializers.swift:398]: Let’s keep them in alphabetical order. AFAICT `TupleExprElement` is the only extension that’s not currently in alphabetical order (I think because it was renamed from `TupleExprElement`). If you want, you can move that as well to restore alphabetical order.

**Review by ahoppen — COMMENTED** [review]
Thank you. Let’s keep the ordering in ConvenienceInitializers.swift alphabetical, though. Or at least not break it any more than it’s already broken.

**Comment by RoBo-Inc** [comment]
Done ✅
All alphabetical now 😃

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftSyntaxBuilder/ListBuilder.swift
Comment (ahoppen) [inline Sources/SwiftSyntaxBuilder/ListBuilder.swift:1]: Ah, sorry that I didn’t catch this as well. You need to add `ListBuilder.swift` to [SwiftSyntaxBuilder/CMakeLists.txt](https://github.com/apple/swift-syntax/blob/main/Sources/SwiftSyntaxBuilder/CMakeLists.txt). We are building swift-syntax using CMake when it gets linked into the compiler.

Comment (RoBo-Inc) [reply Sources/SwiftSyntaxBuilder/ListBuilder.swift:1]: Done ✅
Maybe we should add this info into `Review and CI Testing` section (which already exists) of the `CONTRIBUTING.md` file 🙂

Comment (ahoppen) [reply Sources/SwiftSyntaxBuilder/ListBuilder.swift:1]: Yeah, maybe that would be a good idea 👍🏽

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Review by KidiIT — COMMENTED** [review]
All checked and approved🫶❤️

**Comment by KidiIT** [comment]
All checked out. Thank you and you're welcome🤓❤️

**Comment by ahoppen** [comment]
Thank *you* for the PR.

**Review by KidiIT — COMMENTED** [review]
All reviewed and approved for the job and the application for approval for the new position in September will be completed in the next few days🫶🫶
